x-pack/plugins/fleet/public/hooks/use_agent_version.ts

```diff
@@ -21,12 +21,20 @@ export const useAgentVersion = (): string | undefined => {
     const getVersions = async () => {
       try {
         const res = await sendGetAgentsAvailableVersions();
-        // if the endpoint returns an error, use the fallback versions
-        const versionsList = res?.data?.items ? res.data.items : [kibanaVersion];
+        const availableVersions = res?.data?.items;
 
-        setAgentVersion(versionsList[0]);
+        if (
+          availableVersions &&
+          availableVersions.length > 0 &&
+          availableVersions.indexOf(kibanaVersion) === -1
```

Comment:
@joshdover adding you to reviewers as you worked on https://github.com/elastic/kibana/pull/166150/files#diff-de1953cdeee69d71565b65a14751a0dd602ca5e3c8f81b4854113e6b97cc231b 
Does this logic make sense? To my understanding, we need to check if Kibana version is in available versions to support older ESS versions. Or each version of Kibana have json with available versions generated in a way, that it only contains the current patch version as the first element. Eg. in the future when the latest major is 8.13.0 and say there is a 8.11.2 or smth released, will json contain versions lower then 8.11.2 or it actually will include all 8.12.x and 8.13.x?

Reply:
From my understanding the JSON generated during the build will include all versions released up to that point in time, and nothing later.

On ESS/Stateful, we also add the current Kibana version which is already handled server side. I don't think this change here is needed, but I may not be understanding fully. Could you explain exactly the scenario that isn't currently working?

In any case, if a change is needed, please make it on the server side in https://github.com/elastic/kibana/blob/main/x-pack/plugins/fleet/server/services/agents/versions.ts so that it can be applied everywhere (there are some other backend pieces that consume this logic).

Reply:
I think this test case https://github.com/elastic/kibana/pull/166198/files#diff-949758c47e64dee7bf348fa03e9146d8b87a71a62f37bee09db163c46569675cR22 captures the scenario. Without the change, the returned version would be 8.9.0 while to my understanding we want an earlier version to be picked up.

If I'm not missing anything about this scenario, the change on the server side would be to filter out all the versions greater than the current Kibana version and I guess it will break other places in fleet when we need released agent versions greater than the current Kibana version

Reply:
The current [logic](https://github.com/elastic/kibana/pull/166150/files#diff-ec3e6dccf8414322e363e6177e96902ec8d78829a3e3f5ca8f92d5d989716404) only returns released agent versions in serverless, so I don't think you need any change to fix this for serverless.

Reply:
@juliaElastic the thing I struggle to understand is how we deal with the patch version to the older version of the stack. Eg. when there are 8.12.0, 8.11.0 released but the user updates to say 8.9.3 which was released. The list of released versions will contain 8.12.0 is the first element and this version then will be picked up for all the artifact links in the Fleet UI. So in 8.9.3 kibana the links will point to 8.12.0 which is released, but may not be compatible with this old version. Do we see this as an issue?

Reply:
I think it shouldn't cause an issue if the agent version is newer than kibana (due to version negotiation of the API) and we plan to support it, however I'm not aware if anyone tested this, so there might be some issues to fix. cc @jsoriano @jlind23 to confirm.

Reply:
Yeah, I don't know if this is being tested, maybe from the Agent side? cc @cmacknz

Reply:
Fleet server won't let agents that are newer than Fleet server's version connect, there is a version constrain enforced as part of the check in. 

https://github.com/elastic/fleet-server/blob/d1d155959b0cc9ab2111c5b3dbdab27b5820245c/internal/pkg/api/userAgent.go#L32-L41

We plan to remove this constraint eventually but this is still the case today and for all previously released versions.

Reply:
Then if read @cmacknz's comment correctly we do actually need to make sure that we pick the agent version that matches the Kibana version if this version exists in the list of released agent versions. If we just keep it as it is now (picking the latest released agent version from the list no matter of the Kibana version) it will cause agents not to be able to connect to fleet servers

Reply:
If kibana version is not in the list of agent versions, should the logic return the latest agent version that is <= kibana version? It's not a likely scenario, though might happen if kibana and agent uses different versions eventually.

Reply:
@juliaElastic good point, I changed the logic to accommodate for that and added a test case